pkg/uniqueid/module.go

```diff
@@ -0,0 +1,123 @@
+)
+
+type uniqueIdGenerator struct {
+	kernel.EssentialBackgroundModule
+	kernel.EssentialStage
+
+	logger     log.Logger
+	registerer InstanceRegisterer
+	generator  Generator
+	instance   Instance
+}
+
+var um = struct {
+	sync.Mutex
+	instance *uniqueIdGenerator
+}{}
+
+func ProvideGenerator(ctx context.Context, config cfg.Config, logger log.Logger) (Generator, error) {
+	return provideGenerator(ctx, config, logger)
+}
+
+func ProvideModule() kernel.ModuleFactory {
+	return func(ctx context.Context, config cfg.Config, logger log.Logger) (kernel.Module, error) {
+		return provideGenerator(ctx, config, logger)
+	}
+}
+
+func provideGenerator(ctx context.Context, config cfg.Config, logger log.Logger) (*uniqueIdGenerator, error) {
+	um.Lock()
+	defer um.Unlock()
+
+	if um.instance != nil {
+		return um.instance, nil
+	}
+
+	var err error
+	if um.instance, err = NewGenerator(ctx, config, logger); err != nil {
+		return nil, err
+	}
+
+	return um.instance, nil
+}
+
+func NewGenerator(ctx context.Context, config cfg.Config, logger log.Logger) (*uniqueIdGenerator, error) {
+	settings := Settings{}
+	config.UnmarshalKey("unique_id", &settings)
+
+	logger = logger.WithChannel("unique-id")
+
+	registerer, err := NewInstanceRegisterer(ctx, config, logger, settings)
+	if err != nil {
+		return nil, fmt.Errorf("could not create instance registerer: %w", err)
+	}
+
+	// pass application context once available
+	instance, err := registerer.Register(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("could not register instance: %w", err)
+	}
+
+	logger = logger.WithFields(log.Fields{
+		"machine_id": instance.MachineId,
+	})
+
+	logger.Info("registered instance")
+
+	generator := generatorProvider(instance.MachineId, settings.StartTime)
+	if generator == nil {
+		return nil, fmt.Errorf("could not create sonyflake for instance %d", instance.MachineId)
+	}
+
+	return NewUniqueIdWithInterfaces(logger, registerer, generator, *instance), nil
+}
+
+func NewUniqueIdWithInterfaces(logger log.Logger, registerer InstanceRegisterer, generator Generator, instance Instance) *uniqueIdGenerator {
+	return &uniqueIdGenerator{
+		logger:     logger,
+		registerer: registerer,
+		generator:  generator,
+		instance:   instance,
+	}
+}
+
+func (m *uniqueIdGenerator) Run(ctx context.Context) error {
+	<-ctx.Done()
+
+	// prevent further generations
+	m.generator = nil
+
+	// need a new context here, because the run context is already finished
+	ctx, cancelFn := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFn()
+
+	err := m.registerer.Unregister(ctx, m.instance)
+	if err != nil {
+		return fmt.Errorf("could not unregister instance %d: %w", m.instance.MachineId, err)
+	}
+
+	m.logger.Info("unregistered instance")
+
+	return nil
+}
+
+func (m *uniqueIdGenerator) NextId() (*int64, error) {
+	generator := m.generator
+
+	if generator == nil {
+		return nil, fmt.Errorf("generator not initialized, you need to use the module")
```

Comment:
In what situation does this happen when I forget to use the module? `Run` only sets it to nil after it is done, if I don't use the module, doesn't it just work (but fail to unregister at the end)?